File: rs/core/src/ray_crossing_counter.rs
//! Counts the number of segments crossed by a horizontal ray extending to the
//! right from a given point, in an incremental fashion. This can be used to
//! determine whether a point lies in a polygonal geometry. The class determines
//! the situation where the point lies exactly on a segment. When being used for
//! Point-In-Polygon determination, this case allows short-circuiting the
//! evaluation.
//!
//! This implementation uses the extended-precision orientation test, to provide
//! maximum robustness and consistency within other algorithms.
//!
//! @jts RayCrossingCounter

use geo_types::Coord;

use crate::location::{BOUNDARY, EXTERIOR, INTERIOR};
use crate::orientation::{COLLINEAR, LEFT, index};

pub(crate) struct RayCrossingCounter {
    p: Coord<f64>,
    crossing_count: i32,
    /// true if the test point lies on an input segment
    is_point_on_segment: bool,
}

// The exact coordinate comparisons below are Java's ordinate equality. The
// algorithm's correctness depends on exact vertex identity, so an epsilon must
// never be substituted — if `clippy::float_cmp` is ever enabled, allow it here
// rather than loosening it.
impl RayCrossingCounter {
    /// Determines the location of a point in a ring. This method is an exemplar
    /// of how to use this struct.
    ///
    /// @jts RayCrossingCounter#locatePointInRing(Coordinate,Coordinate[])
    /// @jts-adapter RayCrossingCounter#locatePointInRing(Coordinate,CoordinateSequence)
    ///   — the ports have no sequence abstraction, so only the slice overload is
    ///   ported and it stands in for both. The name carries every parameter type
    ///   because the overload-suffix rule decides the form from the whole Java overload set,
    ///   not from the subset that is ported.
    pub(crate) fn locate_point_in_ring_coordinate_coordinates(
        p: Coord<f64>,
        ring: &[Coord<f64>],
    ) -> i32 {
        let mut counter = RayCrossingCounter::new(p);

        for i in 1..ring.len() {
            let p1 = ring[i];
            let p2 = ring[i - 1];
            counter.count_segment(p1, p2);
            if counter.is_on_segment() {
                return counter.get_location();
            }
        }
        counter.get_location()
    }

    /// @jts RayCrossingCounter#RayCrossingCounter(Coordinate)
    pub(crate) fn new(p: Coord<f64>) -> Self {
        Self {
            p,
            crossing_count: 0,
            is_point_on_segment: false,
        }
    }

    /// Counts a segment.
    ///
    /// @jts RayCrossingCounter#countSegment(Coordinate,Coordinate)
    pub(crate) fn count_segment(&mut self, p1: Coord<f64>, p2: Coord<f64>) {
        // For each segment, check if it crosses a horizontal ray running from
        // the test point in the positive x direction.

        // check if the segment is strictly to the left of the test point
        if p1.x < self.p.x && p2.x < self.p.x {
            return;
        }

        // check if the point is equal to the current ring vertex
        if self.p.x == p2.x && self.p.y == p2.y {
            self.is_point_on_segment = true;
            return;
        }
        // For horizontal segments, check if the point is on the segment.
        // Otherwise, horizontal segments are not counted.
        if p1.y == self.p.y && p2.y == self.p.y {
            let mut minx = p1.x;
            let mut maxx = p2.x;
            if minx > maxx {
                minx = p2.x;
                maxx = p1.x;
            }
            if self.p.x >= minx && self.p.x <= maxx {
                self.is_point_on_segment = true;
            }
            return;
        }
        // Evaluate all non-horizontal segments which cross a horizontal ray to
        // the right of the test pt. To avoid double-counting shared vertices, we
        // use the convention that
        //  - an upward edge includes its starting endpoint, and excludes its final endpoint
        //  - a downward edge excludes its starting endpoint, and includes its final endpoint
        if ((p1.y > self.p.y) && (p2.y <= self.p.y)) || ((p2.y > self.p.y) && (p1.y <= self.p.y)) {
            let mut orient = index(p1, p2, self.p);
            if orient == COLLINEAR {
                self.is_point_on_segment = true;
                return;
            }
            // Re-orient the result if needed to ensure effective segment direction is upwards
            if p2.y < p1.y {
                orient = -orient;
            }
            // The upward segment crosses the ray if the test point lies to the left (CCW) of the segment.
            if orient == LEFT {
                self.crossing_count += 1;
            }
        }
    }

    /// Gets the count of crossings.
    ///
    /// Nothing in the ported subset calls this: `locate_point_in_ring_*` reads
    /// `get_location` instead. It is ported because `pin.json` scopes
    /// `RayCrossingCounter`'s `portedMembers` to the whole file.
    ///
    /// @jts RayCrossingCounter#getCount()
    #[allow(dead_code)]
    pub(crate) fn get_count(&self) -> i32 {
        self.crossing_count
    }

    /// Reports whether the point lies exactly on one of the supplied segments.
    ///
    /// @jts RayCrossingCounter#isOnSegment()
    pub(crate) fn is_on_segment(&self) -> bool {
        self.is_point_on_segment
    }

    /// Gets the location of the point relative to the ring, polygon or
    /// multipolygon from which the processed segments were provided.
    ///
    /// @jts RayCrossingCounter#getLocation()
    pub(crate) fn get_location(&self) -> i32 {
        if self.is_point_on_segment {
            return BOUNDARY;
        }

        // The point is in the interior of the ring if the number of X-crossings is odd.
        if (self.crossing_count % 2) == 1 {
            return INTERIOR;
        }
        EXTERIOR
    }

    /// Tests whether the point lies in or on the ring, polygon or multipolygon
    /// from which the processed segments were provided.
    ///
    /// Nothing in the ported subset calls this, for the same reason as
    /// `get_count`.
    ///
    /// @jts RayCrossingCounter#isPointInPolygon()
    #[allow(dead_code)]
    pub(crate) fn is_point_in_polygon(&self) -> bool {
        self.get_location() != EXTERIOR
    }
}

File: js/src/algorithm/RayCrossingCounter.ts
import type { Coordinate } from "../GeometryAdapter";

import { BOUNDARY, EXTERIOR, INTERIOR } from "../geom/Location";
import { COLLINEAR, LEFT, index } from "./Orientation";

/**
 * Counts the number of segments crossed by a horizontal ray extending to the
 * right from a given point, in an incremental fashion. This can be used to
 * determine whether a point lies in a polygonal geometry. The class determines
 * the situation where the point lies exactly on a segment. When being used for
 * Point-In-Polygon determination, this case allows short-circuiting the
 * evaluation.
 *
 * This class handles polygonal geometries with any number of shells and holes.
 * The orientation of the shell and hole rings is unimportant.
 *
 * This implementation uses the extended-precision orientation test, to provide
 * maximum robustness and consistency within other algorithms.
 *
 * @jts RayCrossingCounter
 */
export class RayCrossingCounter {
  /**
   * Determines the Location of a point in a ring. This method is an exemplar of
   * how to use this class.
   *
   * @param p the point to test
   * @param ring an array of Coordinates forming a ring
   * @return the location of the point in the ring
   * @jts RayCrossingCounter#locatePointInRing(Coordinate,Coordinate[])
   * @jts-adapter RayCrossingCounter#locatePointInRing(Coordinate,CoordinateSequence)
   *   — the ports have no sequence abstraction, so only the array overload is
   *   ported and it stands in for both. The name carries every parameter type
   *   because the overload-suffix rule decides the form from the whole Java overload set, not
   *   from the subset that is ported.
   */
  static locatePointInRingCoordinateCoordinates(p: Coordinate, ring: Coordinate[]): number {
    const counter = new RayCrossingCounter(p);

    for (let i = 1; i < ring.length; i++) {
      const p1 = ring[i];
      const p2 = ring[i - 1];
      counter.countSegment(p1, p2);
      if (counter.isOnSegment()) return counter.getLocation();
    }
    return counter.getLocation();
  }

  private p: Coordinate;
  private crossingCount = 0;
  // true if the test point lies on an input segment
  private isPointOnSegment = false;

  /** @jts RayCrossingCounter#RayCrossingCounter(Coordinate) */
  constructor(p: Coordinate) {
    this.p = p;
  }

  /**
   * Counts a segment.
   *
   * @param p1 an endpoint of the segment
   * @param p2 another endpoint of the segment
   * @jts RayCrossingCounter#countSegment(Coordinate,Coordinate)
   */
  countSegment(p1: Coordinate, p2: Coordinate): void {
    /*
     * For each segment, check if it crosses a horizontal ray running from the
     * test point in the positive x direction.
     */

    // check if the segment is strictly to the left of the test point
    if (p1[0] < this.p[0] && p2[0] < this.p[0]) return;

    // check if the point is equal to the current ring vertex
    if (this.p[0] === p2[0] && this.p[1] === p2[1]) {
      this.isPointOnSegment = true;
      return;
    }
    /*
     * For horizontal segments, check if the point is on the segment.
     * Otherwise, horizontal segments are not counted.
     */
    if (p1[1] === this.p[1] && p2[1] === this.p[1]) {
      let minx = p1[0];
      let maxx = p2[0];
      if (minx > maxx) {
        minx = p2[0];
        maxx = p1[0];
      }
      if (this.p[0] >= minx && this.p[0] <= maxx) {
        this.isPointOnSegment = true;
      }
      return;
    }
    /*
     * Evaluate all non-horizontal segments which cross a horizontal ray to the
     * right of the test pt. To avoid double-counting shared vertices, we use the
     * convention that
     *  - an upward edge includes its starting endpoint, and excludes its final endpoint
     *  - a downward edge excludes its starting endpoint, and includes its final endpoint
     */
    if ((p1[1] > this.p[1] && p2[1] <= this.p[1]) || (p2[1] > this.p[1] && p1[1] <= this.p[1])) {
      let orient = index(p1, p2, this.p);
      if (orient === COLLINEAR) {
        this.isPointOnSegment = true;
        return;
      }
      // Re-orient the result if needed to ensure effective segment direction is upwards
      if (p2[1] < p1[1]) {
        orient = -orient;
      }
      // The upward segment crosses the ray if the test point lies to the left (CCW) of the segment.
      if (orient === LEFT) {
        this.crossingCount++;
      }
    }
  }

  /**
   * Gets the count of crossings.
   *
   * @return the crossing count
   * @jts RayCrossingCounter#getCount()
   */
  getCount(): number {
    return this.crossingCount;
  }

  /**
   * Reports whether the point lies exactly on one of the supplied segments.
   * This method may be called at any time as segments are processed. If the
   * result of this method is true, no further segments need be supplied, since
   * the result will never change again.
   *
   * @return true if the point lies exactly on a segment
   * @jts RayCrossingCounter#isOnSegment()
   */
  isOnSegment(): boolean {
    return this.isPointOnSegment;
  }

  /**
   * Gets the Location of the point relative to the ring, polygon or
   * multipolygon from which the processed segments were provided.
   *
   * This method only determines the correct location if all relevant segments
   * have been processed.
   *
   * @return the Location of the point
   * @jts RayCrossingCounter#getLocation()
   */
  getLocation(): number {
    if (this.isPointOnSegment) return BOUNDARY;

    // The point is in the interior of the ring if the number of X-crossings is odd.
    if (this.crossingCount % 2 === 1) {
      return INTERIOR;
    }
    return EXTERIOR;
  }

  /**
   * Tests whether the point lies in or on the ring, polygon or multipolygon from
   * which the processed segments were provided.
   *
   * This method only determines the correct location if all relevant segments
   * have been processed.
   *
   * @return true if the point lies in or on the supplied polygon
   * @jts RayCrossingCounter#isPointInPolygon()
   */
  isPointInPolygon(): boolean {
    return this.getLocation() !== EXTERIOR;
  }
}
